File: frontend/src/lib/memory/preferenceEvolution.ts
import { getDriver } from './neo4j';

export class PreferenceEvolution {
  
  // Check if user already has a preference for this item
  static async getExistingPreference(userId: string, itemName: string, itemColor?: string): Promise<{
    relationshipType: string;
    nodeId: string;
    relationshipId?: string;
  } | null> {
    const driver = getDriver();
    const session = driver.session();
    
    try {
      // Look for existing preferences for this item (with or without color)
      const query = `
        MATCH (u:User {userId: $userId})-[r]->(n)
        WHERE n.name = $itemName
        ${itemColor ? 'AND n.color = $itemColor' : ''}
        RETURN type(r) as relType, id(n) as nodeId, id(r) as relId
      `;
      
      const result = await session.run(query, {
        userId,
        itemName,
        ...(itemColor && { itemColor })
      });
      
      if (result.records.length > 0) {
        const record = result.records[0];
        return {
          relationshipType: record.get('relType'),
          nodeId: record.get('nodeId').toString(),
          relationshipId: record.get('relId').toString()
        };
      }
      
      return null;
    } finally {
      await session.close();
    }
  }
  
  // Update or replace existing preference
  static async updatePreference(
    userId: string, 
    itemName: string, 
    newPreferenceType: 'PREFERS' | 'AVOIDS_DUE_TO',
    itemColor?: string,
    properties?: any
  ): Promise<boolean> {
    const existingPref = await this.getExistingPreference(userId, itemName, itemColor);
    
    if (!existingPref) {
      // No existing preference, this is handled by normal creation
      return false;
    }
    
    // If same preference type, just update properties
    if (existingPref.relationshipType === newPreferenceType) {
      return await this.updateRelationshipProperties(existingPref.relationshipId!, properties);
    }
    
    // Different preference type - delete old and create new
    await this.deleteRelationship(existingPref.relationshipId!);
    return true; // Signal that we deleted old preference
  }
  
  // Delete a specific relationship
  static async deleteRelationship(relationshipId: string): Promise<boolean> {
    const driver = getDriver();
    const session = driver.session();
    
    try {
      const result = await session.run(`
        MATCH ()-[r]->()
        WHERE id(r) = $relId
        DELETE r
      `, { relId: parseInt(relationshipId) });
      
      console.log(`🗑️ Deleted old preference relationship: ${relationshipId}`);
      return true;
    } catch (error) {
      console.error('Error deleting relationship:', error);
      return false;
    } finally {
      await session.close();
    }
  }
  
  // Update relationship properties
  static async updateRelationshipProperties(relationshipId: string, properties: any): Promise<boolean> {
    const driver = getDriver();
    const session = driver.session();
    
    try {
      await session.run(`
        MATCH ()-[r]->()
        WHERE id(r) = $relId
        SET r += $props, r.lastUpdated = datetime()
      `, { 
        relId: parseInt(relationshipId),
        props: properties
      });
      
      console.log(`🔄 Updated preference relationship: ${relationshipId}`);
      return true;
    } catch (error) {
      console.error('Error updating relationship:', error);
      return false;
    } finally {
      await session.close();
    }
  }
  
  // Clean up orphaned nodes (nodes with no relationships)
  static async cleanupOrphanedNodes(): Promise<void> {
    const driver = getDriver();
    const session = driver.session();
    
    try {
      await session.run(`
        MATCH (n)
        WHERE NOT (n)--() AND NOT n:User
        DELETE n
      `);
      
      console.log(`🧹 Cleaned up orphaned nodes`);
    } catch (error) {
      console.error('Error cleaning up orphaned nodes:', error);
    } finally {
      await session.close();
    }
  }
}

File: backend/multi_agent_system/memory/preference_evolution.py
from typing import Optional, Dict, Any
from .neo4j import get_driver

class PreferenceEvolution:
    
    @staticmethod
    def get_existing_preference(user_id: str, item_name: str, item_color: Optional[str] = None) -> Optional[Dict[str, str]]:
        """Check if user already has a preference for this item"""
        driver = get_driver()
        
        with driver.session() as session:
            # Look for existing preferences for this item (with or without color)
            color_condition = "AND n.color = $itemColor" if item_color else ""
            query = f"""
                MATCH (u:User {{userId: $userId}})-[r]->(n)
                WHERE n.name = $itemName
                {color_condition}
                RETURN type(r) as relType, id(n) as nodeId, id(r) as relId
            """
            
            params = {
                'userId': user_id,
                'itemName': item_name
            }
            if item_color:
                params['itemColor'] = item_color
            
            result = session.run(query, params)
            record = result.single()
            
            if record:
                return {
                    'relationshipType': record['relType'],
                    'nodeId': str(record['nodeId']),
                    'relationshipId': str(record['relId'])
                }
            
            return None
    
    @staticmethod
    def update_preference(
        user_id: str, 
        item_name: str, 
        new_preference_type: str,
        item_color: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None
    ) -> bool:
        """Update or replace existing preference"""
        existing_pref = PreferenceEvolution.get_existing_preference(user_id, item_name, item_color)
        
        if not existing_pref:
            # No existing preference, this is handled by normal creation
            return False
        
        # If same preference type, just update properties
        if existing_pref['relationshipType'] == new_preference_type:
            return PreferenceEvolution.update_relationship_properties(
                existing_pref['relationshipId'], 
                properties or {}
            )
        
        # Different preference type - delete old and create new
        PreferenceEvolution.delete_relationship(existing_pref['relationshipId'])
        return True  # Signal that we deleted old preference
    
    @staticmethod
    def delete_relationship(relationship_id: str) -> bool:
        """Delete a specific relationship"""
        driver = get_driver()
        
        with driver.session() as session:
            try:
                session.run(
                    """
                    MATCH ()-[r]->()
                    WHERE id(r) = $relId
                    DELETE r
                    """, 
                    relId=int(relationship_id)
                )
                
                print(f'🗑️ Deleted old preference relationship: {relationship_id}')
                return True
            except Exception as error:
                print(f'Error deleting relationship: {error}')
                return False
    
    @staticmethod
    def update_relationship_properties(relationship_id: str, properties: Dict[str, Any]) -> bool:
        """Update relationship properties"""
        driver = get_driver()
        
        with driver.session() as session:
            try:
                session.run(
                    """
                    MATCH ()-[r]->()
                    WHERE id(r) = $relId
                    SET r += $props, r.lastUpdated = datetime()
                    """, 
                    relId=int(relationship_id),
                    props=properties
                )
                
                print(f'🔄 Updated preference relationship: {relationship_id}')
                return True
            except Exception as error:
                print(f'Error updating relationship: {error}')
                return False
    
    @staticmethod
    def cleanup_orphaned_nodes() -> None:
        """Clean up orphaned nodes (nodes with no relationships)"""
        driver = get_driver()
        
        with driver.session() as session:
            try:
                session.run(
                    """
                    MATCH (n)
                    WHERE NOT (n)--() AND NOT n:User
                    DELETE n
                    """
                )
                
                print('🧹 Cleaned up orphaned nodes')
            except Exception as error:
                print(f'Error cleaning up orphaned nodes: {error}')